### PR TITLE
REST API: Fix module settings updating

### LINKED ISSUE #6539: REST API: Module settings are not being updated
#### Steps to reproduce the issue
* Open Jetpack and activate Custom Content Types module
* Enable Portfolio Projects

#### What I expected
* To see that Portfolio Projects checkbox in Settings->Writing is checked
* To be able to add new portfolio.

#### What happened instead
* I see a success message upon save: "Updated Custom Content Types settings."
* The post type is actually not enabled.
* Clicking on "Configure Portfolios" button takes me to "Invalid Post Type" page.

This is the same for all modules that have sub options. It can be tested with "Infinite Scroll" and "Markdown" modules for example.

It appears that this issue has been introduced in 8f584fdfe1d795cd8b7e18f7788fc7394bf81957 where @roccotripaldi changed the param filtering logic to support older PHP versions. It seems that when @roccotripaldi was refactoring this, he used `is_string` to filter parameters, but `array_filter` filters the values, and not the keys as it was initially supposed to. This PR changes the logic to filter the keys, and not the values.

cc @roccotripaldi @tyxla

### PR DESCRIPTION
**Pull request by Stoyan0v**
Fixes #6539

#### Changes proposed in this Pull Request:

* Add new method that filters unused parameters.
* Replace the old validation method with the new one.

#### Testing instructions:

* Open Jetpack and activate Custom Content Types module
* Enable Portfolio Projects
* Make sure that post type has been registered.

cc @roccotripaldi @tyxla

### CODE REVIEW

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 1]
```diff
@@ -848,6 +848,26 @@ public function can_request( $request ) {
 			return current_user_can( 'jetpack_configure_modules' );
 		}
 	}
+
+	/**
+	 * Check if unused variables are included in the request.
+	 *
+	 * @since 4.8.0
```
Comment (jeherve) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Could you change this to 4.7.0? It's definitely a blocker, that we would need to include in 4.7.0.

Thanks!

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php [hunk 2]
```diff
@@ -437,8 +437,8 @@ public function update_data( $request ) {
 		// If $params was set via `get_body_params()` there may be some additional variables in the request that can
 		// cause validation to fail. This method verifies that each param was in fact updated and will throw a `some_updated`
 		// error if unused variables are included in the request.
-		$params = array_filter( $params, 'is_string' );
-		unset( $params['context'], $params['slug'] );
+		$filtered = array_filter( array_keys( $params ), array( $this, 'filter_unused_request_params' ) );
+		$params = array_intersect_key( $params, array_flip( $filtered ) );
```
Comment (nylen) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: There is probably a simpler way to do this...

```php
foreach ( array_keys( $params ) as $key ) {
    if ( is_int( $key ) || 'slug' === $key || 'context' === $key ) {
        unset( $params[ $key ] );
    }
}
```

**Comment by nylen** [comment]
Up to the Jetpack team what to do with this feedback, but similarly to https://github.com/Automattic/jetpack/pull/6006#pullrequestreview-20748222 I would prefer not to add an extra method for this one-off task.  This parameter parsing code is a bit weird and really shouldn't be reused anywhere else, either.

File: _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php (hunk 2 above)
Comment (zinigor) [inline _inc/lib/core-api/class.jetpack-core-api-module-endpoints.php]: Wasn't it like that before this change? I'm not sure how `unset` works related to array keys.

**Comment by roccotripaldi** [comment]
Thanks so much @Stoyan0v - LGTM!

>Could you change this to 4.7.0?
@jeherve - what do you mean by that?

**Comment by zinigor** [comment]
I think @jeherve meant that we need to get this fix into `branch-4.7`. We can do it manually after merging to master.

**Comment by zinigor** [comment]
Ported to `branch-4.7` in 1071343.

**Pull request merged**